src/intelligence_layer/core/model.py

```diff
@@ -418,11 +422,40 @@ def to_instruct_prompt(
             input: Any context necessary to solve the task, such as the text to be summarize
             response_prefix: Optional argument to append a string to the beginning of the
                 final agent message to steer the generation
+            input_controls: TextControls for the input part of the prompt
+            instruction_controls: TextControls for the instruction part of the prompt
         """
-        return self.INSTRUCTION_PROMPT_TEMPLATE.to_rich_prompt(
+        rich_prompt = self.INSTRUCTION_PROMPT_TEMPLATE.to_rich_prompt(
             instruction=instruction, input=input, response_prefix=response_prefix
         )
 
+        text_controls = []
+        if instruction_controls:
+            instruction_start = rich_prompt.ranges.get("instruction")[0].start.position  # type: ignore
+            for control in instruction_controls:
+                if control.start + control.length > len(instruction):
+                    raise ValueError(
+                        f"TextControl is out of bounds for instruction {instruction}"
+                    )
+                text_controls.append(
+                    replace(control, start=control.start + instruction_start)
+                )
+
+        if input_controls and input:
+            input_start = rich_prompt.ranges.get("input")[0].start.position  # type: ignore
```

Comment:
What about asserting that both the "input" and "instruction" ranges both are length 1? this makes sure that we don't have any unexpected behavior

Reply:
Good point. Done.